repair/repair.cc

```diff
@@ -2374,6 +2374,67 @@ future<> repair_service::repair_tablets(repair_uniq_id rid, sstring keyspace_nam
     auto task = co_await _repair_module->make_and_start_task<repair::tablet_repair_task_impl>({}, rid, keyspace_name, table_names, streaming::stream_reason::repair, std::move(task_metas), ranges_parallelism);
 }
 
+// It is called by the repair_tablet rpc verb to repair the given tablet
+future<> repair_service::repair_tablet(locator::tablet_metadata_guard& guard, locator::global_tablet_id gid) {
+    auto id = _repair_module->new_repair_uniq_id();
+    rlogger.debug("repair[{}]: Starting tablet repair global_tablet_id={}", id.uuid(), gid);
+    auto& db = get_db().local();
+    auto table_id = gid.table;
+    auto tablet_id = gid.tablet;
+
+    auto host2ip = [&addr_map = _addr_map] (locator::host_id host) -> future<gms::inet_address> {
+        auto ip = addr_map.local().find(raft::server_id(host.uuid()));
+        if (!ip) {
+            throw std::runtime_error(format("Could not get ip address for host {} from raft_address_map", host));
+        }
+        co_return *ip;
+    };
+
+    lw_shared_ptr<replica::table> t;
+    try {
+        t = db.find_column_family(table_id).shared_from_this();
+    } catch (replica::no_such_column_family& e) {
+        co_return;
+    }
+    auto& tmap = guard.get_tablet_map();
+    auto s = t->schema();
+    auto keyspace_name = s->ks_name();
+    auto table_name = s->cf_name();
+    std::vector<sstring> table_names = {table_name};
+    auto myhostid = guard.get_token_metadata()->get_my_id();
+
+    auto range = tmap.get_token_range(tablet_id);
+    auto& info = tmap.get_tablet_info(tablet_id);
+    std::vector<gms::inet_address> nodes;
+    std::vector<shard_id> shards;
+    shard_id master_shard_id;
+    for (auto& r : info.replicas) {
+        auto shard = r.shard;
+        auto ip = co_await host2ip(r.host);
```

Comment:
I think we avoid playing the tricks to filter excluded replica nodes everywhere. What we can do is to skip the tablet that has a replica node that is excluded in the tablet repair scheduler itself. The already scheduled repair will fail in case a node is excluded after the scheduling anyway. 

Later, when the excluded node is removed from the replica set. Repair will be scheduled again and complete successfully. 

Note, we need to repair all replicas to advance the repair time for the purpose of tombstone gc.

Reply:
It's better to not delay repair until removenode is finished, it may take a long time.

Also, there was this idea to invoke repair to speed up replica rebuild, in which case we will have to support excluded nodes.

What's wrong with my suggestion? We already do it for migration.

Reply:
Because we could not use the partial repair for tombstone gc purpose. It is better to wait for the replica set of a tablet is in normal and health state before we run regular repair.  Also, we should make the removenode a higher priority over regular repairs tasks since the cluster is in a degraded state. 

Running repair for rebuild task is a different operation, we probably will not mix it with the regular repair operations, similar to RBNO and regular repair.

Reply:
> Because we could not use the partial repair for tombstone gc purpose.

We can, excluded nodes are not coming back. It's not like with the old repair where exclusion is user-provided and temporary. Exclusion in topology is permanent. So we can consider the whole tablet as repaired even if you ignore excluded nodes.

Reply:
We already rely on excluded nodes not coming back in migration and rebuild, where we don't do topology barrier on excluded nodes.

Reply:
> Also, we should make the removenode a higher priority over regular repairs tasks since the cluster is in a degraded state.

The scheduler should already do that, right? It schedules rebuilding migrations before scheduling repair.

Reply:
Btw, if a node is excluded, when do we remove the node from the tablet replicas list?

Reply:
> But if a round of the scheduling does not finish all rebuild tasks, the repair tasks could still be picked before the remaining rebuild tasks, no?

Currently, I think the main plan making will try to schedule as much rebuilds as allowed by the limits. That won't make room for repairs.

But theoretically, the two could share the load. Perhaps they even should. Rebuilds could be more important than repair, but not always. Repair may be triggered to avoid running out of disk space due to tombstones.

It's best to allow repair to be run with excluded nodes, and leave the decision to the scheduler what's more important.

@avikivity Opinion?

Reply:
> Btw, if a node is excluded, when do we remove the node from the tablet replicas list?

After admin invokes removenode or replace, tablet scheduler will schedule rebuilding transitions for tablets which have replicas on excluded nodes. Uses the same algorithm as during decommission. Replica set is changed when rebuilding transition completes successfully.

Reply:
When a node is marked as excluded, why don't we remove the node from the tablet replica set as soon as we mark it as excluded.  The node is not going back anyway. It is pointless to leave it in the replica set and filter it everywhere.

Reply:
> Currently, I think the main plan making will try to schedule as much rebuilds as allowed by the limits. That won't make room for repairs.

> But theoretically, the two could share the load. Perhaps they even should. Rebuilds could be more important than repair, but not always. Repair may be triggered to avoid running out of disk space due to tombstones.

Such emergency repair should be scheduled much early explicitly when we have the feedback of the tombstone gc ratio as part of the scheduling decision, not use it as a last minute resort hoping rebuild task give repair some room to do it. 

> It's best to allow repair to be run with excluded nodes, and leave the decision to the scheduler what's more important.

Again, I think we should not do the optimization which is not a clear win now. Let's make the basic working first.

Reply:
> When a node is marked as excluded, why don't we remove the node from the tablet replica set as soon as we mark it as excluded. The node is not going back anyway. It is pointless to leave it in the replica set and filter it everywhere.

One of the reasons is view pairing.

If B is excluded in {A, B, C}, we want to change the replica set to {A, D, C}, not {A, C} then {A, C, D}. That would change the replica pairing between base and view, which is based on the index in the replica set

It's not like we filter it everywhere. request coordinator doesn't filter it.

Reply:
> Again, I think we should not do the optimization which is not a clear win now. Let's make the basic working first.

I'm not saying we should do the optimization now. 

We should fix the bug which makes repair hang when there are excluded nodes. And the right way to fix it is to do what migration does, because it's consistent with the rest of the system, and allows flexibility in scheduler logic should we need it later.

Reply:
> One of the reasons is view pairing.

Another reason is that we don't want to change replica sets (prev, next) concurrently with transitions, which may assume replica set stability. Doing so would create complexity in transition implementation which would have to interpret sudden replica set changes. Currently they would get confused about `leaving` or `pending` disappearing.

Reply:
> > Again, I think we should not do the optimization which is not a clear win now. Let's make the basic working first.
> 
> I'm not saying we should do the optimization now.

Good. 
 
> We should fix the bug which makes repair hang when there are excluded nodes. And the right way to fix it is to do what migration does, because it's consistent with the rest of the system, and allows flexibility in scheduler logic should we need it later.

We do not have the hang issue with excluded nodes anymore with the latest version.
- scheduler will stop scheduling tablet with exclude nodes
- previously scheduled repair will fail because the excludes nodes will cause the repair to fail